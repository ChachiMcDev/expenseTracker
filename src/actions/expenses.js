import { v4 as uuid } from 'uuid'
import database from '../firebase/firebase';

//Expenses actions

//component calls action generator
//action generator returns an object
//component dispatches object to redux
//redux store changes

//*** asych flow     */
//component calls action generator
//action generator returns a function
//component dispatches function ((?) =>   redux does not natively allow functions, redux middlware is used ) 
//////// using redux-thunk as the middleware
//function runs (has the abilit to dispatch other actions and do whaterver it wants)




//ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});


export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt }

        return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};


//REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`)
            .remove()
            .then(() => {
                dispatch(removeExpense({ id }))
            });
    };
};





//EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id, updates));
        });
    };
};



//SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});


export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses`)
            .once('value')
            .then((snapshot) => {
                const expenses = [];
                snapshot.forEach((childsnapshot) => {
                    expenses.push({
                        id: childsnapshot.key,
                        ...childsnapshot.val()
                    })
                });

                dispatch(setExpenses(expenses));
            });
    };

};


