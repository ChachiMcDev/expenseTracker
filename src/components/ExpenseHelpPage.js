import React from 'react';

const ExpenseHelpPage = ()=> (
    <div>this is the expense help page</div>
)

export default ExpenseHelpPage;