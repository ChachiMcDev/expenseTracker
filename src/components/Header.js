import React from 'react';
import {NavLink} from 'react-router-dom';

const Header = () => (
    <header>
        <h1>Expensify It!</h1>


        <NavLink  
            exact to="/" 
            activeClassName="is-active"
            activeStyle={{
                color: "blue"
            }}
        >
            Dashboard
        </NavLink>

        <NavLink 
            exact to="/create" 
            activeClassName="is-active"
            activeStyle={{
                color: "blue"
            }}
        >
            Create Expense
        </NavLink>

        <NavLink 
            exact to="/help" 
            activeClassName="is-active"
            activeStyle={{
                color: "red"
            }}
            >
            Need Help?
        </NavLink>

    </header>
)

export default Header;