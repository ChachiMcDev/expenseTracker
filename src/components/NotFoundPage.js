import React from 'react';
import { Link} from 'react-router-dom';

const NotFoundPage = ()=> (
    <div>
        error... error... error... <Link to="/">Go Home</Link>
    </div>
)

export default NotFoundPage;