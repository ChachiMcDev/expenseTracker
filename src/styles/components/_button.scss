.button {
  background: $blue;
  border: none;
  color: white;
  display: inline-block;
  font-size: $font-size-large;
  font-weight: 300;
  line-height: 1;
  padding: $s-size;
  text-decoration: none;
}

.button--link {
  background: none;
}

.button--secondary{
  background: #888;
}

.button--glogin {
  width: 189px;
  margin-bottom: 15px;
}