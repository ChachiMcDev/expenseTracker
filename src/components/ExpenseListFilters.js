
import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';
import { setTextFilter, setSortByDateFilter,  setSortByAmountFilter, setStartDate, setEndDate } from '../actions/filters';




export class ExpenseListFilters extends React.Component {

    state = {
        calendarFocused: null
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
      }

    onDatesChange = ({ startDate, endDate }) => {  
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onTextChange = (event) => {
        this.props.setTextFilter(event.target.value);
    }

    onSortChange = (e) => {
        if (e.target.value === 'date') {
          this.props.setSortByDateFilter();
              
        } else if (e.target.value === 'amount') {
          this.props.setSortByAmountFilter();
            
        }
      };

    render() {
        return (
            <div>
                <input
                    type="text"
                    value={this.props.filters.text}
                    onChange={this.onTextChange} 
                />

                <select
                    name="sortby"
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange }
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                            showClearDates={true}
                    
                        />
            </div>
        )
    }

}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) =>({
    
        setTextFilter:  (text) => dispatch(setTextFilter(text)),
        setSortByDateFilter: () => dispatch(setSortByDateFilter()),
        setSortByAmountFilter: () => dispatch(setSortByAmountFilter()),
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate))

});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);




//import React from 'react';
//import { connect } from 'react-redux';
//import { DateRangePicker } from 'react-dates';
//import { setTextFilter, setSortByDateFilter, setSortByAmountFilter, setStartDate, setEndDate } from '../actions/filters';
//
//
//const ExpenseListFilters = (props) => (
//    <div>
//        <input 
//            type="text" 
//            value={props.filters.text} 
//            onChange={(event) => {
//            props.dispatch(setTextFilter(event.target.value));
//        }}/>
//        
//        <select 
//            name="sortby"
//            value={props.filters.sortBy}
//            onChange={(event)=>{
//                let targ = event.target.value;
//                if(targ === "amount"){
//                    props.dispatch(setSortByAmountFilter())
//                }
//                if(targ === "date"){
//                    props.dispatch(setSortByDateFilter())
//                }
//        }}>
//            <option value="date">Date</option>
//            <option value="amount">Amount</option>
//            <option value="start-to-end-date">
//            <DateRangePicker 
//                startDate={props.filters.startDate}
//                endDate={props.filters.endDate}
//                onDatesChange={}
//                focusedInput={}
//                onFocusChange={}
//            />
//                Date: start to end</option>
//        </select>
//    </div>
//);
//
//
//
//
//export default connect(mapStateToProps)(ExpenseListFilters);