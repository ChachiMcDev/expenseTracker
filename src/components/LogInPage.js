import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';


const fbreturns = () => {
    return 'facebook'
}

export const LogInPage = ({ startLogin }) => (
    <div className="box-layout">
        
        <div className="box-layout__box">
            <h1 className="box-layout__title">Expense Tracker App</h1>
            <p>Do you know where your expenses are right now?</p>
            <button className="button" onClick={startLogin}>Login with Google</button>
            <button className="button" loginwith={fbreturns()} onClick={startLogin}>Login with Facebook</button>
        </div>      
    </div>
);



const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
})

const mapStateToProps = (state, props) => {

    return {
        loginwith: () => {fbreturns()}
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(LogInPage);