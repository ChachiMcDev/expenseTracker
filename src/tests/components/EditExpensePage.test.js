import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expensesData';


let editExpenseSpy, removeExpenseSpy, historySpy, wrapper;

beforeEach(() =>{
    editExpenseSpy = jest.fn();
    removeExpenseSpy = jest.fn();
    historySpy = { push: jest.fn() };
    wrapper = shallow(<EditExpensePage
                         editExpense={editExpenseSpy} 
                         removeExpense={removeExpenseSpy} 
                         history={historySpy} 
                         expense={expenses[1]} 
                    />);
});


test('should render add expense page without errors', ()=>{
    expect(wrapper).toMatchSnapshot();
});


test('should handle editExpense', ()=>{
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(historySpy.push).toHaveBeenLastCalledWith('/');
    expect(editExpenseSpy).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
    
});

test('should handle removeExpense', ()=>{
    wrapper.find('button').simulate('click');
    expect(historySpy.push).toHaveBeenLastCalledWith('/');
    expect(removeExpenseSpy).toHaveBeenLastCalledWith({id: expenses[1].id});
})


 