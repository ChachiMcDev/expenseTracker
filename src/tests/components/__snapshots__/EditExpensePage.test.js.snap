// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render add expense page without errors 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 295,
        "createdAt": -345600000,
        "description": "rent",
        "id": "1",
        "note": "this is a note on ID 1",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
